### DISCUSSION TITLE
How to interpret results in YOLOv8 exported to TFJS or ONNX?

### DISCUSSION THREAD

**Discussion by Anto5040** [ultralytics/ultralytics#6205] [Q&A]
Hello,
I trained a YOLOv8 model on custom data so it can detect only one class. It works wonderfully on python, but I tried to export it to TFJS and have quite some problems. They were resolved by downgrading tensorflow:
`pip install tensorflow==2.13.1`

With that, I could finally export my model to TFJS. After some tries, I managed to get a prediction for an image as a JSON file. The problem is that the prediction is a [1,5,8400] dimension tensor. As far as I have been able to research, this means in practice [x_center, y_center, width, heigth, confidence in class 1] for each of the 8400 predictions made.

This seems quite strange because in the JSON there are a whole lot (hundreds) of predictions with scores equal to or really close to one, which shouldn't be there. I have tried the same model in python in it's native format and with the same image, I only get 7 or 8 predictions with scores higher than 0.5. 

After that, I have tried the same thing but converting the model to ONNX, and that resulted in a JSON with 42000 values. That maps to the 1x5x8400 tensor that I got earlier, so at least it is consistent.

What am I missing?

EDIT: The values with the ONNX format are just concatenated. I'd higly recommend this blogpost to understand the format:
https://dev.to/andreygermanov/how-to-create-yolov8-based-object-detection-web-service-using-python-julia-nodejs-javascript-go-and-rust-4o8e#process_the_output_nodejs

Thanks in advance.
Am I missing something?

**Comment by Anto5040** [comment]
Thanks to the tutorial provided by @AndreyGermanov here https://dev.to/andreygermanov/how-to-create-yolov8-based-object-detection-web-service-using-python-julia-nodejs-javascript-go-and-rust-4o8e#process_the_output_nodejs I have managed to get a result.
But that result wildly differs from the one I get on python because I get more than 50 different object detections, even after NMS, with scores of 1 or really close to that. 

Is there someone that has had the same issue?

**Reply by AndreyGermanov** [reply]
Interesting case. Could you provide the models (both Pytorch and ONNX) and the image that you used? I will try them and let you know what's up and what to do.

**Reply by Anto5040** [reply]
Thank you for answering @AndreyGermanov 
I have managed to find my error. I trained my model with a small amount of images, so that could have been one of the problems.

Furthermore, I had pretty bad detections with the already trained models, but that was because of my preprocessing.
I had the following preprocessing for the image:
```

function readImage(path) {
  const buf = fs.readFileSync(path);
  const img = new Image();
  img.src = buf; // Asigna directamente el buffer leído a src
  return img;
}

function imageToInput(image, numChannels) {
  // Tamaño esperado por el modelo
  const targetWidth = 640;
  const targetHeight = 640;

  // Crear un canvas para manejar el redimensionamiento de la imagen
  const canvas = createCanvas(targetWidth, targetHeight);
  const ctx = canvas.getContext('2d');

  // Dibujar la imagen en el canvas con el tamaño objetivo
  ctx.drawImage(image, 0, 0, targetWidth, targetHeight);
  const resizedImageData = ctx.getImageData(0, 0, targetWidth, targetHeight).data;

  const size = targetWidth * targetHeight * numChannels;
  const inputTensor = new Float32Array(size);

  for (let i = 0; i < size; i += numChannels) {
    inputTensor[i] = resizedImageData[i] / 255;
    inputTensor[i + 1] = resizedImageData[i + 1] / 255;
    inputTensor[i + 2] = resizedImageData[i + 2] / 255;
  }

  return new ort.Tensor('float32', inputTensor, [1, numChannels, targetHeight, targetWidth]);
}

```
But using yours and editing a bit, I have managed to get plausible results with this code:
```
async function prepare_input(path) {
  const buf = fs.readFileSync(path);
  const img = sharp(buf);
  const md = await img.metadata();
  const [img_width, img_height] = [md.width, md.height];
  const pixels = await img.removeAlpha()
      .resize({width: 640, height: 640, fit: 'fill'})
      .raw()
      .toBuffer();
  const red = [], green = [], blue = [];
  for (let index = 0; index < pixels.length; index += 3) {
      red.push(pixels[index] / 255.0);
      green.push(pixels[index + 1] / 255.0);
      blue.push(pixels[index + 2] / 255.0);
  }
  const input = [...red, ...green, ...blue];
  return new ort.Tensor('float32', Float32Array.from(input), [1, 3, 640, 640]);
}

```

I hope this helps others with a similar problem.
Again, thank you Andrey for your post, because it was incredibly helpful!

**Reply by H-Tudor** [reply]
Original Answer: https://github.com/orgs/ultralytics/discussions/2354#discussioncomment-7915608

I had a similar issue, and from my research i gathered the following:

Theory:
- yolov5 has an output of shape (batchSize, 25200, 85) (Num classes + box[x,y,w,h] + confidence[c])
- yolov8 has an output of shape (batchSize, 84, 8400) (Num classes + box[x,y,w,h])

My yolo model has an output of 1x7x8400:
- 1 = batch size for single image inference
- 7 = 3 classes + 4 params (x, y, w, h)
- 8400 = subdivisions I suppose (must be concatenated using NMS)

C++ decoder can be found at: ultralytics/examples/YOLOv8-CPP-Inference/inference.cpp/Inference::runInference
I encourage reading these examples and also some modification of the .cpp and .h is required

**Reply by pderrenger** [reply]
@H-Tudor it's great to hear that you've resolved your issue and that you found the resources provided by the community helpful. Indeed, training with a larger and more diverse dataset can significantly improve the model's performance and generalization.

Regarding the preprocessing, it's crucial to ensure that the image preprocessing steps match those expected by the model. The adjustments you made to your preprocessing code seem to have aligned with the model's requirements, leading to better results.

For future reference, when working with YOLOv8 or any other model, always ensure that the input preprocessing in your deployment environment matches the preprocessing used during training. This includes image resizing, normalization, and channel ordering.

If you encounter any further issues or have questions about YOLOv8, feel free to reach out. The community is here to support you. Good luck with your continued work on object detection! 😊🚀
